Automattic/bugomattic
# Add error handling and improved validation

#### What Does This PR Add/Change?

This PR adds some overdue error handling and validation to the app!

It adds...
- A top level error boundary to catch any errors that bubble up in the app
- Handling for when the reporting config fails to load
- Validation of the reporting config as it is being parsed
- Deeper validation (like checking if a feature ID even exists) of state coming in from URL history

Not done in this PR:
We need to update our logic for handling the currently active step. I will be reworking that in another PR! Just worth noting that you could set a bad URL and end up in a weird UI state. 

##### Validation Philosophy

There are lots of tradeoffs and decisions when you add validation and error handling to a site! I want to lay out the general tenets that I followed for transparency...

- **Validate all user input before saving to the store** -- The app should be able to trust what is in the store.  If a state update comes in any way from user input, validate it in the reducer. The main vector for this right now is URL manipulation. 
- **Trust programmatic state updates** -- in contrast to the above, if any piece of state is directly calculated and set by the code, don't do any fancy validation or fallback. If something breaks here, it should break big and loud, because it means there's a bug in our code! Our global error handler will catch the error and notify us. 
- **Reserve the global error handler for unexpected behavior** -- If something falls in the range of expected user interaction, we should handle it gracefully and not throw. Adding fallbacks for URL handling is a good example of this.
-  **Throw liberally for bad reporting configs** -- consider this the one exception to the above! An incorrectly formatted reporting config JSON document just opens up a whole can of worms we don't want. Let's keep on top of that JSON document aggressively!

#### Testing Instructions

I've gone through and tested a lot of permutations of these cases already! There's also a lot of unit/component/integration testing for these changes added in this PR.

If you'd like to play around, some examples of what you can look at...
- Navigating forward and back through the flow and make sure everything works as it did before.
- Manipulate the query params in the URL to set bad data. You may end up in a weird UI state (see note above about reworking the active step logic), but the app should never crash!
- Manipulate local reporting config you are using. You should see the global error display and see an error logged in the console.

#### Issues



Related to #  
Closes #
